server/src/main/java/org/elasticsearch/index/mapper/flattened/FlattenedFieldMapper.java

```diff
@@ -708,7 +716,57 @@ public IndexFieldData.Builder fielddataBuilder(FieldDataContext fieldDataContext
 
         @Override
         public ValueFetcher valueFetcher(SearchExecutionContext context, String format) {
-            return SourceValueFetcher.identity(name(), context, format);
+            if (format != null) {
+                throw new IllegalArgumentException("Field [" + name() + "] of type [" + typeName() + "] doesn't support formats.");
+            }
+            return sourceValueFetcher(context.isSourceEnabled() ? context.sourcePath(name()) : Collections.emptySet());
+        }
+
+        private SourceValueFetcher sourceValueFetcher(Set<String> sourcePaths) {
+            return new SourceValueFetcher(sourcePaths, null) {
+                @Override
+                @SuppressWarnings("unchecked")
+                protected Object parseSourceValue(Object value) {
+                    if (value instanceof Map<?, ?> valueAsMap && valueAsMap.isEmpty() == false) {
+                        final Map<String, List<String>> result = filterIgnoredValues((Map<String, Object>) valueAsMap);
+                        return result.isEmpty() ? null : result;
+                    }
+                    if (value instanceof String valueAsString && valueAsString.length() < ignoreAbove) {
+                        return valueAsString;
+                    }
+                    return null;
+                }
+
+                private Map<String, List<String>> filterIgnoredValues(final Map<String, Object> values) {
+                    final Map<String, List<String>> result = new HashMap<>();
+                    for (final Map.Entry<String, Object> entry : values.entrySet()) {
+                        final List<String> validValues = filterIgnoredValues(entry.getValue());
+
+                        if (validValues.isEmpty() == false) {
+                            final String flattenedFieldName = entry.getKey();
+                            result.put(flattenedFieldName, validValues);
+                        }
+                    }
+                    return result;
+                }
+
+                private List<String> filterIgnoredValues(final Object entryValue) {
+                    final List<String> validValues = new ArrayList<>();
+
+                    if (entryValue instanceof List<?> valueAsList) {
+                        for (Object value : valueAsList) {
+                            if (value instanceof String valueAsString && valueAsString.length() < ignoreAbove) {
```

Comment:
What can these be, other than String? Shall we assert?

Reply:
What is that you want to assert? I still need to do the cast anyway...

Reply:
In that example, I see that we have a number so it's fine to ignore. I wonder if there are cases where we can have mixed singletons (e.g. strings) with maps pointing to more strings that can be ignored.

Reply:
I have added more tests covering cases I could imagine...I don't see anything missing.